src/main/java/com/google/api/codegen/configgen/MessageGenerator.java

```diff
@@ -0,0 +1,253 @@
+/* Copyright 2017 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.api.codegen.configgen;
+
+import com.google.api.codegen.configgen.nodes.ConfigNode;
+import com.google.api.codegen.configgen.nodes.FieldConfigNode;
+import com.google.api.codegen.configgen.nodes.ListItemConfigNode;
+import com.google.api.codegen.configgen.nodes.ScalarConfigNode;
+import com.google.api.tools.framework.util.VisitsBefore;
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.Message;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Generates a Message from a ConfigNode representation. */
+public class MessageGenerator extends NodeVisitor {
+  private static final String TYPE_KEY = "type";
+
+  private Message.Builder messageBuilder;
+  private final FieldDescriptor field;
+  private List<Object> values;
+
+  public MessageGenerator(Message.Builder messageBuilder) {
+    this(messageBuilder, null);
+  }
+
+  private MessageGenerator(Message.Builder messageBuilder, FieldDescriptor field) {
+    this.messageBuilder = messageBuilder;
+    this.field = field;
+  }
+
+  @VisitsBefore
+  boolean generate(FieldConfigNode node) {
```

Comment:
optional: seems a little strange for a public class to have mostly package-private methods